### PR TITLE
wire in configuration for subpartitions

### PR DESCRIPTION
**Pull request by agavra**
next PR will be dedicated to integration testing - this allows us to "unblock" the release

### CODE REVIEW

File: kafka-client/src/main/java/dev/responsive/kafka/config/ResponsiveConfig.java [hunk 1]
```diff
@@ -190,4 +211,57 @@ public String toString() {
...
+      final int kafkaPartitions,
+      final int desiredNum,
+      final TableName name,
+      final String changelogTopicName
+  ) {
+    final int factor = (desiredNum == NO_SUBPARTITIONS) ? 1 : desiredNum / kafkaPartitions;
+    final int computedRemoteNum = factor * kafkaPartitions;
+
+    if (actualRemoteCount.isPresent() && actualRemoteCount.getAsInt() != computedRemoteNum) {
+      throw new IllegalArgumentException(String.format("%s was configured to %d, which "
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/config/ResponsiveConfig.java:254]: nit: should we use ConfigException?

File: kafka-client/src/main/java/dev/responsive/kafka/config/ResponsiveConfig.java [hunk 2]
```diff
@@ -190,4 +211,57 @@ public String toString() {
...
+    return getSubPartitioner(
+        actualRemoteCount,
+        kafkaPartitions,
+        getInt(STORAGE_DESIRED_NUM_PARTITION_CONFIG),
+        name,
+        changelogTopicName
+    );
+  }
+
+  static SubPartitioner getSubPartitioner(
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/config/ResponsiveConfig.java:243]: This should be private, right? If it's only visible for testing, we can just go through the public method above with a mock admin right?

(Or make this a static util method in the `SubPartitioner` class instead)

File: kafka-client/src/main/java/dev/responsive/kafka/config/ResponsiveConfig.java [hunk 3]
```diff
@@ -190,4 +211,57 @@ public String toString() {
...
+    final int computedRemoteNum = factor * kafkaPartitions;
+
+    if (actualRemoteCount.isPresent() && actualRemoteCount.getAsInt() != computedRemoteNum) {
+      throw new IllegalArgumentException(String.format("%s was configured to %d, which "
+              + "given %s partitions in kafka topic %s would result in %d remote partitions "
+              + "for table %s (remote partitions must be a multiple of the kafka partitions). "
+              + "The remote store is already initialized with %d partitions - it is backwards "
+              + "incompatible to change this. Please set %s to %d.",
+          STORAGE_DESIRED_NUM_PARTITION_CONFIG, desiredNum, kafkaPartitions,
+          changelogTopicName, computedRemoteNum, name.cassandraName(),
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/config/ResponsiveConfig.java:260]: super nit: prefer the kafkaName in any user-facing error/logs
```suggestion
          changelogTopicName, computedRemoteNum, name.kafkaName(),
```

File: kafka-client/src/test/java/dev/responsive/utils/SubPartitionerTest.java [hunk 4]
```diff
@@ -57,4 +57,13 @@ public void shouldIterateAllSubPartitions() {
     assertThat(result, contains(4, 5));
   }
 
+  @Test
+  public void shouldNotChangeSalt() {
```
Comment (ableegoldman) [inline kafka-client/src/test/java/dev/responsive/utils/SubPartitionerTest.java:61]: nice 😄

**Review by ableegoldman — APPROVED** [review]
Some nits but on the whole LGTM

File: kafka-client/src/main/java/dev/responsive/utils/SubPartitioner.java [hunk 5]
```diff
@@ -47,14 +50,39 @@ public class SubPartitioner {
...
   private final Function<Bytes, Integer> hasher;
 
+  public static SubPartitioner create(
+      final OptionalInt actualRemoteCount,
+      final int kafkaPartitions,
+      final int desiredNum,
+      final TableName name,
+      final String changelogTopicName
+  ) {
+    final int factor = (desiredNum == ResponsiveConfig.NO_SUBPARTITIONS) ? 1 : desiredNum / kafkaPartitions;
```
Comment (rodesai) [inline kafka-client/src/main/java/dev/responsive/utils/SubPartitioner.java:68]: Should we err on the side of over-partitioning instead? Also, consider including this formula in the doc.

**Review by rodesai — APPROVED** [review]
LGTM to merge

**Pull request merged**